### ISSUE TITLE
Add support for destructured import definitions (esm module imports)

### ISSUE THREAD

**Issue by iambumblehead** [sinonjs/sinon#2361]
Add support for destructured definitions in esm-module-only environments.

The requested feature is to support this style of import
```javascript
import { stub } from 'sinon';
```

Currently, an esm-module-only environment must import sinon as shown,
```javascript
import sinon from 'sinon';

sinon.stub();
```

**Comment by fatso83** [comment]
OK, what is the use case? Purely ergonomics? Sinon is a test utility so there is no bundling performance improvement to be had and doing member imports makes it even less obvious that all the fakes and stubs created belong to the Sinon sandbox.

**Comment by iambumblehead** [comment]
I only wanted to let you know sinon does not support destructured imports. Personally, I prefer the supported way of importing one default namespace, but am aware of those who prefer destructured imports.

All core modules and most public packages (I've used) support _both_ default and destructured import assignments.

A few days ago, I removed babel from an application using destructured sinon imports. Those no longer worked using the un-transpiled sources and it was necessary to update each one to import the default sinon namespace.

**Comment by fatso83** [comment]
It is a bit weird that this should not work, seeing that #1833 was supposed to do exactly that. I'll have a quick look.

**Comment by fatso83** [comment]
OK, I see now that this does not work without modifications in the standard Node case:
```javascript
$ cat test.mjs
import { spy } from "sinon";

console.log(spy);

$ node test.mjs
file:///tmp/test1/test.mjs:1
import { spy } from "sinon";
         ^^^
SyntaxError: Named export 'spy' not found. The requested module 'sinon' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'sinon';
const { spy } = pkg;
```

Explicitly importing the esm  export file and changing the file ending "fixes" it:
```javascript
$ cat test.mjs
import { spy } from "sinon/pkg/sinon-esm.mjs";

console.log(spy);

$ cp node_modules/sinon/pkg/sinon-esm.js  node_modules/sinon/pkg/sinon-esm.mjs
$ node test.mjs
[Function: spy]
```

Reading up [on the Node docs](https://nodejs.org/api/packages.html#packages_type) now. Also found [an interesting article](https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html) that seems to cover the various other use cases (Webpack, browsers, etc) for hybrid modules.

**Comment by fatso83** [comment]
OK, so #1835 implemented support for this, but our ESM bundle is only usable by _bundlers_ like Webpack and Rollup, that use the [unofficial `pkg.module` field](https://stackoverflow.com/a/47537198/200987) to find the entry point for an ESM. _Normal Node_ in ESM mode (meaning, being started from a file ending in `.mjs`) does not use this field, and so will regard `sinon` as a CommonJS module unless _something else_ is present.

The linked article in my previous comment (about hybrid modules using a single source tree) basically summed up what needed to be done:
> Author your code in ES6, ES-Next or Typescript using import and export.
>
>From this base, you can import either ES modules or CommonJS modules using import. The reverse is not true. If you author in CommonJS you cannot easily consume ES modules.

OK, easily done, since we [officially support writing ES2017 syntax](https://github.com/sinonjs/sinon/blob/master/COMPATIBILITY.md) these days, but still a bit of work (feel free, though!) to convert the entire source tree. _BUT_ ... this is assuming we are not willing to have a split tree of some kind - which we already do! We already produce an ESM for bundlers, so it should be possible to just point to that using [conditional exports](https://nodejs.org/api/packages.html#packages_conditional_exports). I'll have a quick go!

**Comment by fatso83** [comment]
Adding this essentially made it work from the consumer side of things:
```
+  "exports": {
+    "require": "./pkg/sinon.cjs",
+    "import": "./pkg/sinon-esm.js"
+  },
+  "type": "module",
```

but ... then we could no longer run our tests. So I do see there was a lot of stuff written further down on the conditional exports documentation on this, and some of it might be possible to get going, but I will leave that for someone else that can be bothered to invest the time 😓 

To me, the easiest route forward seems to be to simply make this project use ESM all the way, set `type: "module"` and possibly be done with it (still producing the bundles in `pkg/`, though, if there is any interest in that). I am not interested enough myself to do this, but someone who really wants it can put aside an hour (or more?) to fix the imports and possible side effects.

**Comment by mroderick** [comment]
Before anyone sets out to convert this entire codebase to ESM, I suggest they read this three part series:

* https://gils-blog.tayar.org/posts/using-jsm-esm-in-nodejs-a-practical-guide-part-1/
* https://gils-blog.tayar.org/posts/using-jsm-esm-in-nodejs-a-practical-guide-part-2/
* https://gils-blog.tayar.org/posts/using-jsm-esm-in-nodejs-a-practical-guide-part-3/

Especially part 2, where the author explains dual-mode libraries.

TL;DR, we would have to use a transpiler to create CommonJS from ESM sources.

I'm not saying that it can't be done or that it shouldn't be done ... I'm cautioning people that it will be a considerable undertaking that will require a fair amount of testing to ensure everything works in browsers, nodejs and popular post processors like `webpack`, `rollup`, `esbuild`, etc.

I would suggest making a solution proposal in a separate issue, so we can iron out the details on how to go about it and how to ensure that end users are unaffected by the changes.

**Comment by iambumblehead** [comment]
I'm a little shy recommending this package I wrote, but its been working well for me https://www.npmjs.com/package/esmock

It is an alternative to proxyquire and testdouble.

I develop an enterprisey application with a lot of modules and haven't had any problems with this.

**Comment by mroderick** [comment]
> I'm a little shy recommending this package I wrote, but its been working well for me https://www.npmjs.com/package/esmock
> 
> It is an alternative to proxyquire and testdouble.
> 
> I develop an enterprisey application with a lot of modules and haven't had any problems with this.

I'll take it for a spin in a work project. Thanks for the tip

**Comment by mroderick** [comment]
One more thing that needs investigating: `sinon` uses `this` internally to keep track of some things. Will that still work when consuming it with ESM?

```js
import { stub } from 'sinon';
```

**Comment by fatso83** [comment]
Should be closed by #2382 and the great work done by @perrin4869, whose work was left to dry for way too long.
